File: hogvm/python/stl/print.py
import re

# Copied from clickhouse_driver.util.escape, adapted only from single quotes to backquotes.
escape_chars_map = {
    "\b": "\\b",
    "\f": "\\f",
    "\r": "\\r",
    "\n": "\\n",
    "\t": "\\t",
    "\0": "\\0",
    "\a": "\\a",
    "\v": "\\v",
    "\\": "\\\\",
}
singlequote_escape_chars_map = {**escape_chars_map, "'": "\\'"}
backquote_escape_chars_map = {**escape_chars_map, "`": "\\`"}


# Copied from clickhouse_driver.util.escape_param
def escape_string(value: str) -> str:
    return "'{}'".format("".join(singlequote_escape_chars_map.get(c, c) for c in str(value)))


# Copied from clickhouse_driver.util.escape, adapted from single quotes to backquotes. Added a $.
def escape_identifier(identifier: str | int) -> str:
    if isinstance(identifier, int):  # In HogQL we allow integers as identifiers to access array elements
        return str(identifier)
    # HogQL allows dollars in the identifier.
    if re.match(r"^[A-Za-z_$][A-Za-z0-9_$]*$", identifier):
        return identifier
    return "`{}`".format("".join(backquote_escape_chars_map.get(c, c) for c in identifier))


def print_hog_value(obj):
    if isinstance(obj, list):
        return f"[{', '.join(map(print_hog_value, obj))}]"
    if isinstance(obj, dict):
        return f"{{{', '.join([f'{print_hog_value(key)}: {print_hog_value(value)}' for key, value in obj.items()])}}}"
    if isinstance(obj, tuple):
        if len(obj) < 2:
            return f"tuple({', '.join(map(print_hog_value, obj))})"
        return f"({', '.join(map(print_hog_value, obj))})"
    if obj is True:
        return "true"
    if obj is False:
        return "false"
    if obj is None:
        return "null"
    if isinstance(obj, str):
        return escape_string(obj)
    return str(obj)


def print_hog_string_output(obj):
    if isinstance(obj, str):
        return str(obj)
    return print_hog_value(obj)

File: hogvm/typescript/src/stl/print.ts
import { isHogCallable, isHogClosure, isHogDate, isHogDateTime, isHogError } from '../objects'

const escapeCharsMap: Record<string, string> = {
    '\b': '\\b',
    '\f': '\\f',
    '\r': '\\r',
    '\n': '\\n',
    '\t': '\\t',
    '\0': '\\0',
    '\v': '\\v',
    '\\': '\\\\',
}

const singlequoteEscapeCharsMap: Record<string, string> = {
    ...escapeCharsMap,
    "'": "\\'",
}

const backquoteEscapeCharsMap: Record<string, string> = {
    ...escapeCharsMap,
    '`': '\\`',
}

export function escapeString(value: string): string {
    return `'${value
        .split('')
        .map((c) => singlequoteEscapeCharsMap[c] || c)
        .join('')}'`
}

export function escapeIdentifier(identifier: string | number): string {
    if (typeof identifier === 'number') {
        return identifier.toString()
    }
    if (/^[A-Za-z_$][A-Za-z0-9_$]*$/.test(identifier)) {
        return identifier
    }
    return `\`${identifier
        .split('')
        .map((c) => backquoteEscapeCharsMap[c] || c)
        .join('')}\``
}

export function printHogValue(obj: any, marked: Set<any> | undefined = undefined): string {
    if (!marked) {
        marked = new Set()
    }
    if (typeof obj === 'object' && obj !== null && obj !== undefined) {
        if (
            marked.has(obj) &&
            !isHogDateTime(obj) &&
            !isHogDate(obj) &&
            !isHogError(obj) &&
            !isHogClosure(obj) &&
            !isHogCallable(obj)
        ) {
            return 'null'
        }
        marked.add(obj)
        try {
            if (Array.isArray(obj)) {
                if ((obj as any).__isHogTuple) {
                    if (obj.length < 2) {
                        return `tuple(${obj.map((o) => printHogValue(o, marked)).join(', ')})`
                    }
                    return `(${obj.map((o) => printHogValue(o, marked)).join(', ')})`
                }
                return `[${obj.map((o) => printHogValue(o, marked)).join(', ')}]`
            }
            if (isHogDateTime(obj)) {
                const millis = String(obj.dt)
                return `DateTime(${millis}${millis.includes('.') ? '' : '.0'}, ${escapeString(obj.zone)})`
            }
            if (isHogDate(obj)) {
                return `Date(${obj.year}, ${obj.month}, ${obj.day})`
            }
            if (isHogError(obj)) {
                return `${String(obj.type)}(${escapeString(obj.message)}${
                    obj.payload ? `, ${printHogValue(obj.payload, marked)}` : ''
                })`
            }
            if (isHogClosure(obj)) {
                return printHogValue(obj.callable, marked)
            }
            if (isHogCallable(obj)) {
                return `fn<${escapeIdentifier(obj.name ?? 'lambda')}(${printHogValue(obj.argCount)})>`
            }
            if (obj instanceof Map) {
                return `{${Array.from(obj.entries())
                    .map(([key, value]) => `${printHogValue(key, marked)}: ${printHogValue(value, marked)}`)
                    .join(', ')}}`
            }
            return `{${Object.entries(obj)
                .map(([key, value]) => `${printHogValue(key, marked)}: ${printHogValue(value, marked)}`)
                .join(', ')}}`
        } finally {
            marked.delete(obj)
        }
    } else if (typeof obj === 'boolean') {
        return obj ? 'true' : 'false'
    } else if (obj === null || obj === undefined) {
        return 'null'
    } else if (typeof obj === 'string') {
        return escapeString(obj)
    }
    return obj.toString()
}

export function printHogStringOutput(obj: any): string {
    if (typeof obj === 'string') {
        return obj
    }
    return printHogValue(obj)
}
